### DISCUSSION TITLE
Trying to get iPXE for DeepComputing DC-ROMA RISCV to work

### DISCUSSION THREAD

**Discussion by arcivanov** [ipxe/ipxe#1370] [Q&A]
The board in question is this: https://deepcomputing.io/product/dc-roma-risc-v-mainboard/#specs

The result is after chainloading from Grub is a blank screen. Any suggestions what to try?

Built IPXE EFI RISCV64 as follows:

`make CROSS=riscv64-linux-gnu- bin-riscv64-efi/ipxe.efi DEBUG=init:3`

Configuration changes:

```patch
diff --git a/src/config/console.h b/src/config/console.h
index 0ff328b7..28c39aed 100644
--- a/src/config/console.h
+++ b/src/config/console.h
@@ -35,7 +35,7 @@ FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
  */
 
 //#define      CONSOLE_SERIAL          /* Serial port console */
-//#define      CONSOLE_FRAMEBUFFER     /* Graphical framebuffer console */
+#define        CONSOLE_FRAMEBUFFER     /* Graphical framebuffer console */
 //#define      CONSOLE_SYSLOG          /* Syslog console */
 //#define      CONSOLE_SYSLOGS         /* Encrypted syslog console */
 //#define      CONSOLE_VMWARE          /* VMware logfile console */
diff --git a/src/config/general.h b/src/config/general.h
index 763a34aa..c1df8628 100644
--- a/src/config/general.h
+++ b/src/config/general.h
@@ -116,17 +116,17 @@ FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
 //#define      IMAGE_ELF               /* ELF image support */
 //#define      IMAGE_MULTIBOOT         /* MultiBoot image support */
 //#define      IMAGE_PXE               /* PXE image support */
-//#define      IMAGE_SCRIPT            /* iPXE script image support */
+#define        IMAGE_SCRIPT            /* iPXE script image support */
 //#define      IMAGE_BZIMAGE           /* Linux bzImage image support */
 //#define      IMAGE_COMBOOT           /* SYSLINUX COMBOOT image support */
-//#define      IMAGE_EFI               /* EFI image support */
+#define        IMAGE_EFI               /* EFI image support */
 //#define      IMAGE_SDI               /* SDI image support */
 //#define      IMAGE_PNM               /* PNM image support */
 #define        IMAGE_PNG               /* PNG image support */
 #define        IMAGE_DER               /* DER image support */
 #define        IMAGE_PEM               /* PEM image support */
-//#define      IMAGE_ZLIB              /* ZLIB image support */
-//#define      IMAGE_GZIP              /* GZIP image support */
+#define        IMAGE_ZLIB              /* ZLIB image support */
+#define        IMAGE_GZIP              /* GZIP image support */
 //#define      IMAGE_UCODE             /* Microcode update image support */
 
 /*
@@ -149,29 +149,29 @@ FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );
 #define LOGIN_CMD              /* Login command */
 #define SYNC_CMD               /* Sync command */
 #define SHELL_CMD              /* Shell command */
-//#define NSLOOKUP_CMD         /* DNS resolving command */
-//#define TIME_CMD             /* Time commands */
-//#define DIGEST_CMD           /* Image crypto digest commands */
-//#define LOTEST_CMD           /* Loopback testing commands */
-//#define VLAN_CMD             /* VLAN commands */
+#define NSLOOKUP_CMD           /* DNS resolving command */
+#define TIME_CMD               /* Time commands */
+#define DIGEST_CMD             /* Image crypto digest commands */
+#define LOTEST_CMD             /* Loopback testing commands */
+#define VLAN_CMD               /* VLAN commands */
 //#define PXE_CMD              /* PXE commands */
-//#define REBOOT_CMD           /* Reboot command */
-//#define POWEROFF_CMD         /* Power off command */
-//#define IMAGE_TRUST_CMD      /* Image trust management commands */
-//#define IMAGE_CRYPT_CMD      /* Image encryption management commands */
-//#define PCI_CMD              /* PCI commands */
-//#define PARAM_CMD            /* Request parameter commands */
-//#define NEIGHBOUR_CMD                /* Neighbour management commands */
-//#define PING_CMD             /* Ping command */
-//#define CONSOLE_CMD          /* Console command */
-//#define IPSTAT_CMD           /* IP statistics commands */
-//#define PROFSTAT_CMD         /* Profiling commands */
-//#define NTP_CMD              /* NTP commands */
-//#define CERT_CMD             /* Certificate management commands */
-//#define IMAGE_MEM_CMD                /* Read memory command */
+#define REBOOT_CMD             /* Reboot command */
+#define POWEROFF_CMD           /* Power off command */
+#define IMAGE_TRUST_CMD        /* Image trust management commands */
+#define IMAGE_CRYPT_CMD        /* Image encryption management commands */
+#define PCI_CMD                /* PCI commands */
+#define PARAM_CMD              /* Request parameter commands */
+#define NEIGHBOUR_CMD          /* Neighbour management commands */
+#define PING_CMD               /* Ping command */
+#define CONSOLE_CMD            /* Console command */
+#define IPSTAT_CMD             /* IP statistics commands */
+#define PROFSTAT_CMD           /* Profiling commands */
+#define NTP_CMD                /* NTP commands */
+#define CERT_CMD               /* Certificate management commands */
+#define IMAGE_MEM_CMD          /* Read memory command */
 #define IMAGE_ARCHIVE_CMD      /* Archive image management commands */
 #define SHIM_CMD               /* EFI shim command (or dummy command) */
-//#define USB_CMD              /* USB commands */
+#define USB_CMD                /* USB commands */
 
 /*
  * ROM-specific options
```

grub.cfg:

```patch
--- grub.cfg.bak        2024-12-22 17:50:52.300867831 -0500
+++ grub.cfg    2024-12-22 20:14:32.849563295 -0500
@@ -208,6 +208,15 @@
 # This file provides an easy way to add custom menu entries.  Simply type the
 # menu entries you want to add after this comment.  Be careful not to change
 # the 'exec tail' line above.
+menuentry "iPXE" {
+  load_video
+  insmod gzio
+  insmod part_gpt
+  insmod chain
+  search --no-floppy --fs-uuid --set=root 2c3f8ce4-d531-4b52-abcc-1f9c6d9c2642
+  chainloader /boot/ipxe.efi console config
+}
+
 ### END /etc/grub.d/40_custom ###
 
 ### BEGIN /etc/grub.d/41_custom ###
```

**Comment by NiKiZe** [comment]
A few pointers, don't modify config, many of the ones that you did change doesn't make sense to change, leave everything untill you actually know you do have to change it. But especially when having issues, always revert to non modified configs, and consider using named configs instead.

Have you tried booting from anything other than grub? Grub changes a great deal of the boot process, which not always is compatible with iPXE.

**Reply by arcivanov** [reply]
I have tried booting the unmodified `ipxe.efi` via grub as well. I started messing with the configs only after blank screen, thinking maybe I need video initialization or console to be compiled etc.

I have not tried booting anything other than grub moreover I'm not sure I can. It's an early access platform so documentation is unfortunately very sparse.

There are 4 partitions, with sda1 and sda2 being the bootloaders (bbl and uboot spl), sda3 is efi with a single `bootriscv64.efi` on it. The latter file contains reference to "grub" in strings inside the executable and the EFI partition contains no configuration at all, so I'm not even sure I can boot anything other grub, hence my attempt to chainload ipxe.efi.

**Reply by NiKiZe** [reply]
Can you replace bootriscv64.efi? Or can you boot from a USB device with such a file?

**Reply by arcivanov** [reply]
As in rename ipxe.efi to bootriscv64.efi and replace the existing file? That's a thought! I'll try it tomorrow. Thanks!

**Reply by arcivanov** [reply]
Nope didn't work, the system is stuck on the u-boot splash screen.

![image](https://github.com/user-attachments/assets/142684e1-24d9-4075-8458-dfb40a666dd5)

**Reply by arcivanov** [reply]
And I used the default build configuration too. Any suggestions how to approach this?

There is a file `/boot/vf2_uEnv.txt` which contains this:

```
fdt_high=0xffffffffffffffff
initrd_high=0xffffffffffffffff
kernel_addr_r=0x44000000
kernel_comp_addr_r=0x90000000
kernel_comp_size=0x10000000
fdt_addr_r=0x48000000
ramdisk_addr_r=0x48100000
#Move distro to first boot to speed up booting
boot_targets=distro mmc0 dhcp
#Fix wrong fdtfile name
fdtfile=starfive/jh7110-starfive-visionfive-2-framework.dtb
#Fix missing bootcmd
bootcmd=run load_distro_uenv;run bootcmd_distro
```

I've seen the platform-specific Makefiles change kernel addresses etc depending on what's being built. Any changes I need to make based on the above data?

Also FYI, the kernel  loads devicetree via grub when loading kernel.

**Reply by NiKiZe** [reply]
Try building with DEBUG=efi_init, however the splash screen might hide any text output?

**Reply by arcivanov** [reply]
Nope, still stuck on splash with no text... Any other ideas?

**Reply by arcivanov** [reply]
The internal developer was able to see (through serial console to which I have no access yet) the following:
https://github.com/DC-DeepComputing/Framework/issues/1#issuecomment-2561777620

My guess is this is caused by "console=ttyS0,115200: command not found"?:

```
iPXE 1.21.1+ (g83ba3) -- Open Source Network Boot Firmware -- https://ipxe.org/
Features: DNS HTTP iSCSI TFTP VLAN SRP AoE EFI Menu
console=ttyS0,115200: command not found
INIT shutdown cachedhcp2...
INIT shutdown tcp...
INIT shutdown rbg...
INIT shutdown devices...
INIT shutdown efi_tick...
INIT shutdown heap...
INIT shutdown efi_watchdog...
INIT shutdown complete
Application failed, r = 1
```

**Reply by NiKiZe** [reply]
Yes it looks like iPXE is trying to execute `console=ttyS0,115200` either this is from the command line to ipxe, which i don't think we had support for in efi, or loaded from elsewhere, but the only way that should be supported is thru autoexec.ipxe

Is there any config file or cmdline.txt or similar that might contain anything like it?

Try creating autoexec.ipxe next to the ipxe binary,
```
#!ipxe
imgstat
shell
```

**Reply by arcivanov** [reply]
Still need to sort out the UART - it's not my console dump, but the DC's engineer's

**Reply by NiKiZe** [reply]
You shouldn't need the serial output at all, you should get a proper shell on console (grub is working?), but you could add `dhcp` to the script and instead check the network for traffic.

**Reply by arcivanov** [reply]
> You shouldn't need the serial output at all, you should get a proper shell on console (grub is working?), but you could add `dhcp` to the script and instead check the network for traffic.

That's the first thing I've tried: https://github.com/ipxe/ipxe/discussions/1370#discussion-7732403
I could try again, of course with `imgstat` and `shell` in grub, but the screen was blank with nothing happening via grub - that's what I've started with.

**Reply by NiKiZe** [reply]
A few differences, you did start ipxe thru grub with configuration changes...

Last output did suggest that there is parameters or something else causing ipxe to give "command not found"

I do suggest you run ipxe with embedded script to try and override the command line parameters, and there enforce dhcp.

Can the board be booted from usb flash device? If it can, this would be a way to get the log written to disk. (I'll dig up relevant commit if this is something you can and want to try.)

**Reply by arcivanov** [reply]
The first time I ran IPXE was in Grub, compiled with no customization or command line. But I will try again with vanilla uncustomized build, embedded scripts and grub command line.

I'm not sure in re USB boot, there isn't really a boot select menu due to u-boot.

**Reply by arcivanov** [reply]
@NiKiZe the cause for the console is that U-boot is configured with `bootargs=console=tty1 console=ttyS0,115200  debug rootwait  earlycon=sbiboo` UEFI env var.
Setting it to blank gets IPXE to wait for Ctrl-B!

```
StarFive # env set bootargs ""
StarFive # boot
Scanning mmc 1:3...
Found EFI removable media binary efi/boot/bootriscv64.efi
943104 bytes read in 48 ms (18.7 MiB/s)
EFI handle 0xf76e2b60 systab 0xf7ed9d10
EFI does not provide protocol ConsoleControl
EFI protocol ef9fc172-a1b2-4693-b327-6d32fc416042 is at 0xf7fcf200
EFI does not provide protocol c12b1873-ac17-4176-ac77-7ecb4defffec
EFI does not provide protocol AcpiTable
EFI protocol 8b843e20-8132-4852-90cc-551a4e4a7f1c is at 0xf7fcf100
EFI does not provide protocol 3152bca5-eade-433d-862e-c01cdc291f44
EFI does not provide configuration table eb9d2d30-2d88-11d3-9a16-0090273fc14d
EFI does not provide configuration table f2fd1544-9794-4a2c-992e-e5bbcf20e394
EFI configuration table eb9d2d31-2d88-11d3-9a16-0090273fc14d is at 0xf66ac000
EFI image base address 0xf6377000
EFI image device path /VenHw(e61d73b9-a384-4acc-aeab-82e828f3628b)/SD(1)/SD(1)/HD(3,GPT,fc1c4670-d302-4064-ab03-95f410c3d887,0x4000,0x92000)
iPXE initialising devices...
file:autoexec.ipxe... Not found (https://ipxe.org/7f4de18e)
file:/autoexec.ipxe... Not found (https://ipxe.org/7f4de18e)



iPXE 1.21.1+ (g83ba3) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP iSCSI TFTP VLAN SRP AoE EFI Menu

Press Ctrl-B for the iPXE command line...
```
